In [2]:
from xglass import *
from pprint import pprint
xglass = XGLASS()
import dotenv
dotenv.load_dotenv()
try:
    r = xglass.auth(os.environ['XGLASS_USER'], os.environ['XGLASS_PASS'])
except Exception as e:
    print(e)
    xglass.auth(os.environ['XGLASS_USER'], os.environ['XGLASS_PASS'])
    
try:
    vehicules = xglass.searchVehiculeImmat('ez817mg')
    vehicule = vehicules[0]
    planche = xglass.getPlancheData(vehicule)
    pare_brise = [c for c in planche.xGlassCalques if c.libelle == "PARE-BRISE"][0]

    xglass.getPieces(planche, pare_brise)
    pare_briseElement = xglass.elements_principaux[0]
    for pare_brise_piece_avecAM in pare_briseElement.pieces:
        pprint(pare_brise_piece_avecAM.eurocode)

    pare_brise_piece_avecAM = [piece for piece in pare_briseElement.pieces if piece.eurocode == '861101J110'][0]
    pare_brise_piece_sansAM = [piece for piece in pare_briseElement.pieces if piece.eurocode == '861104P010'][0]

    pare_brise_piece_avecAMPieces = xglass.getPieceAm(pare_briseElement, pare_brise_piece_avecAM)
    pare_brise_piece_sansAMPieces = xglass.getPieceAm(pare_briseElement, pare_brise_piece_sansAM)
except Exception as e:
    print(e)


xglass.close()

Traceback (most recent call last):
  File "_pydevd_bundle/pydevd_cython.pyx", line 563, in _pydevd_bundle.pydevd_cython.PyDBFrame._handle_exception
  File "_pydevd_bundle/pydevd_cython.pyx", line 297, in _pydevd_bundle.pydevd_cython.PyDBFrame.do_wait_suspend
  File "C:\Users\stan\AppData\Roaming\Python\Python310\site-packages\debugpy\_vendored\pydevd\pydevd.py", line 1976, in do_wait_suspend
    keep_suspended = self._do_wait_suspend(thread, frame, event, arg, suspend_type, from_this_thread, frames_tracker)
  File "C:\Users\stan\AppData\Roaming\Python\Python310\site-packages\debugpy\_vendored\pydevd\pydevd.py", line 2011, in _do_wait_suspend
    time.sleep(0.01)
KeyboardInterrupt


name 'getLabel' is not defined
Logout success


In [10]:
planche.__dict__

{'elementsHorsCalque': [{'adas': False,
   'id': '5761',
   'libelle': 'ENJOLIVEUR BRAS ESSUIE- GLACE',
   'piecesAm': [],
   'type': 'C',
   'vitre': False,
   'vitreMobile': False},
  {'adas': False,
   'id': '4639D',
   'libelle': 'SUPPORT PHARE D',
   'piecesAm': [],
   'type': 'C',
   'vitre': False,
   'vitreMobile': False},
  {'adas': False,
   'id': '2426',
   'libelle': 'GICLEUR LAVE- GLACE',
   'piecesAm': [],
   'type': 'C',
   'vitre': False,
   'vitreMobile': False},
  {'adas': False,
   'id': '5059',
   'libelle': 'TUYAU REMPLISSAGE RESERVOIR LAVE- GLACE',
   'piecesAm': [],
   'type': 'C',
   'vitre': False,
   'vitreMobile': False},
  {'adas': False,
   'id': '6232D',
   'libelle': 'CAPUCHON PHARE D',
   'piecesAm': [],
   'type': 'C',
   'vitre': False,
   'vitreMobile': False},
  {'adas': False,
   'id': '1961G',
   'libelle': 'ENJOLIVEUR AV CADRE GLACE PORTE AR G',
   'piecesAm': [],
   'type': 'C',
   'vitre': False,
   'vitreMobile': False},
  {'adas': False,
   'i

In [7]:
from xglass import *
xglass = XGLASS()
r = xglass.auth()
# print(r.cookies)
# r = xglass.get(XGLASS_MAIN_URL)
# print(r.cookies)
XGLASS_searchImmat = f"{XGLASS_URL}/ajax/searchImmat.html"

s = xglass.session.get(XGLASS_searchImmat,
                        # cookies=r.cookies,
    params={
        'selectionsCriteres.immatriculation': 'DS808DZ',
    }
)
print(s.text)

FR17218001S 9913910
Before login
JSESSIONID=6C63B8B191E7ABA381B56188198A29DD
SameSite=None
After login
Session cookies
JSESSIONID=6C63B8B191E7ABA381B56188198A29DD
SameSite=None
Response cookies



















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
<head>
    <title>X'Glass - Le Portail Bris de Glace </title>
	<!-- meta dediee a le redirection lors des appels ajax en etant deconnecte -->
	<meta name="xglass-ajax-redirect" />

	<meta name="description" content="X'Glass est la solution de chiffrage pour les pièces de vitrage et optiques pour les véhicules de type VL et VUL."/>


	<link rel="stylesheet" type="text/css" href="scripts/extjs/resources/css/ext-all.css" />
    
    
    
		<link rel="stylesheet" type="text/css" href="styles/pegase/equipAuto.css" />
	    <link rel="stylesheet" type="text/css" href="/styles/pegase/theme.css" />
	    
	    <script type="text/javascript" src="/scripts/prototype.js"></script>
	    <script type="text/javascript" src="/scripts/s

In [131]:
xglass.session.post(XGLASS_searchImmat,
                        cookies=r.cookies,
    params={
        'selectionsCriteres.immatriculation': 'DS808DZ',
        'idMarque': None,
    }
).text

'\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">\n<head>\n    <title>X\'Glass - Le Portail Bris de Glace </title>\n\t<!-- meta dediee a le redirection lors des appels ajax en etant deconnecte -->\n\t<meta name="xglass-ajax-redirect" />\n\n\t<meta name="description" content="X\'Glass est la solution de chiffrage pour les pièces de vitrage et optiques pour les véhicules de type VL et VUL."/>\n\n\n\t<link rel="stylesheet" type="text/css" href="scripts/extjs/resources/css/ext-all.css" />\n    \n    \n    \n\t\t<link rel="stylesheet" type="text/css" href="styles/pegase/equipAuto.css" />\n\t    <link rel="stylesheet" type="text/css" href="/styles/pegase/theme.css" />\n\t    \n\t    <script type="text/javascript" src="/scripts/prototype.js"></script>\n\t    <script type="text/javascript" src="/scripts/scriptaculous.js"></script>\n\t    <script type="text/javascript" src="/scripts/extjs/adapter/ext/ext-base.js"></script>\n\t\t<script type="text/j

In [118]:
print(s.text)




















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
<head>
    <title>X'Glass - Le Portail Bris de Glace </title>
	<!-- meta dediee a le redirection lors des appels ajax en etant deconnecte -->
	<meta name="xglass-ajax-redirect" />

	<meta name="description" content="X'Glass est la solution de chiffrage pour les pièces de vitrage et optiques pour les véhicules de type VL et VUL."/>


	<link rel="stylesheet" type="text/css" href="scripts/extjs/resources/css/ext-all.css" />
    
    
    
		<link rel="stylesheet" type="text/css" href="styles/pegase/equipAuto.css" />
	    <link rel="stylesheet" type="text/css" href="/styles/pegase/theme.css" />
	    
	    <script type="text/javascript" src="/scripts/prototype.js"></script>
	    <script type="text/javascript" src="/scripts/scriptaculous.js"></script>
	    <script type="text/javascript" src="/scripts/extjs/adapter/ext/ext-base.js"></script>
		<script type="text/javascript" src="/scripts/extjs/ext-all-debug.js"></scr

In [129]:
print(requests.post(XGLASS_searchImmat,
    data={
        'selectionsCriteres.immatriculation': 'DS808DZ',
        # 'idMarque': '',
    },
    cookies=r.cookies,
    headers={
        "accept": "text/html,application/xhtml+xml,application/xml",
        "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36",
        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        "X-Requested-With": "XMLHttpRequest",
        "host": "portail-xglass.com",
        "origin": "https://portail-xglass.com",
    }
).text)

{"actionErrors":[],"actionMessages":[],"criteresVariante":null,"errorCode":"10","errorMessage":null,"errorMessages":[],"errors":{},"fieldErrors":{},"infoMessage":null,"locale":"fr_FR","regroupementVariantes":null,"selectionsCriteres":{"anneeDeCommercialisation":null,"carrosserie":null,"criteres":{},"cylindreeLit":null,"energie":null,"idMarque":null,"idModele":null,"identAdmin":null,"immatriculation":"DS808DZ","portesNbr":null,"puissanceFisc":null,"vitessesBte":null,"vitessesNbr":null},"texts":null,"variantes":null}


In [97]:
XGLASS_modeles = f"{XGLASS_URL}/ajax/modeles.html"
import json
data = {
        "method": "list",
        "marque": 56,
    }
r = requests.get(
    XGLASS_modeles,
    # data={
    #     # 'selectionsCriteres.immatriculation': 'DS808DZ',
    #     # 'idMarque': '',
    #     "method": "list",
    #     "marque": 56,
    # },
    json= json.dumps(data),
    cookies=xglass.session.cookies,
    headers={
        "accept": "*/*",
        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        "Host": "portail-xglass.com",
        "Origin": "https://portail-xglass.com",
        "Pragma": "no-cache",
        "referer": "https://portail-xglass.com/initRechercheVehicule.html",
        # "sec-ch-ua": 'Not A(Brand";v="8", "Chromium";v="132", "Google Chrome";v="132',
        # "sec-ch-ua-mobile": "?0",
        # "sec-ch-ua-platform": "Windows",
        # "sec-fetch-dest": "empty",
        'Sec-Fetch-Dest': 'document',
        "sec-fetch-mode": "cors",
        # "sec-fetch-site": "same-origin",
        'Sec-Fetch-Site': 'none',
        "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 Safari/537.36",
        "X-Requested-With": "XMLHttpRequest",
    },
)
print(r.text)




















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
<head>
    <title>X'Glass - Le Portail Bris de Glace </title>
	<!-- meta dediee a le redirection lors des appels ajax en etant deconnecte -->
	<meta name="xglass-ajax-redirect" />

	<meta name="description" content="X'Glass est la solution de chiffrage pour les pièces de vitrage et optiques pour les véhicules de type VL et VUL."/>


	<link rel="stylesheet" type="text/css" href="scripts/extjs/resources/css/ext-all.css" />
    
    
    
		<link rel="stylesheet" type="text/css" href="styles/pegase/equipAuto.css" />
	    <link rel="stylesheet" type="text/css" href="/styles/pegase/theme.css" />
	    
	    <script type="text/javascript" src="/scripts/prototype.js"></script>
	    <script type="text/javascript" src="/scripts/scriptaculous.js"></script>
	    <script type="text/javascript" src="/scripts/extjs/adapter/ext/ext-base.js"></script>
		<script type="text/javascript" src="/scripts/extjs/ext-all-debug.js"></scr

In [91]:
import pprint
# pprint.pprint(r.request.headers)
for header in r.headers:
    print(header, r.headers[header])
# r.

Date Mon, 10 Feb 2025 10:41:05 GMT
Server Apache
Cache-Control no-cache, no-store, max-age=0, must-revalidate
Pragma no-cache
Expires 0
X-XSS-Protection 1; mode=block
X-Frame-Options ALLOWALL
X-Content-Type-Options nosniff
Content-Type application/json;charset=UTF-8
Content-Language fr-FR
Content-Length 163
Keep-Alive timeout=5, max=100
Connection Keep-Alive


In [15]:
import pandas as pd
# pd.DataFrame(s.json())
from pprint import pprint
pprint(s.json())

JSONDecodeError: Expecting value: line 20 column 1 (char 19)

In [21]:
for header in xglass.session.headers:
    print(header, xglass.session.headers[header])

user-agent Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36
Accept-Encoding gzip, deflate
accept text/html,application/xhtml+xml,application/xml
Connection keep-alive
Content-Type application/x-www-form-urlencoded; charset=UTF-8
X-Requested-With XMLHttpRequest
host portail-xglass.com
origin https://portail-xglass.com
Cookie JSESSIONID=D058667A8A1E4FCD1E1A86DBFF22B66F; SameSite=None


In [36]:
for header in s.headers:
    print(header, s.headers[header])

Date Mon, 10 Feb 2025 09:19:27 GMT
Server Apache
Cache-Control no-cache, no-store, max-age=0, must-revalidate
Pragma no-cache
Expires 0
X-XSS-Protection 1; mode=block
X-Frame-Options ALLOWALL
X-Content-Type-Options nosniff
Content-Type application/json;charset=UTF-8
Content-Language fr-FR
Content-Length 520
Keep-Alive timeout=5, max=100
Connection Keep-Alive


In [207]:
import requests
from xglass import *
import bs4 as bs
from pprint import pprint

session = requests.Session()
headers = {
        "accept": "text/html,application/xhtml+xml,application/xml",
        "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36",
        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        # "Content-Type": "application/json",
        "X-Requested-With": "XMLHttpRequest",
        "Host": "portail-xglass.com",
        "Origin": "https://portail-xglass.com",
    }
session.headers.update(headers)
r = session.get(XGLASS_URL)

JSESSIONID = session.cookies.get_dict().get("JSESSIONID")

r = session.post(
    XGLASS_LOGIN_URL,
    data={
        "j_username": XGLASS_USER,
        "j_password": XGLASS_PASS,
        "spring-security-redirect": "/mainMenu.html",
    },
    verify=False,
    allow_redirects=False,
)

if session.cookies.get_dict().get("JSESSIONID") == JSESSIONID:
    print("Login failed")
    session.close()
    raise Exception("Login failed")




In [208]:
def ensure_logged(r):
    login_form = bs.BeautifulSoup(r.text, "html.parser")
    if login_form.find("form"):
        print("Login failed, login_form found")
        session.close()
        raise Exception("Login failed")

In [209]:
r = session.get(XGLASS_MAIN_URL)
ensure_logged(r)
print(r.text)

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//FR"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">





















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
    <head>
                <!-- HTTP 1.1 -->
        <meta http-equiv="Cache-Control" content="no-store"/>
        <!-- HTTP 1.0 -->
        <meta http-equiv="Pragma" content="no-cache"/>
        <!-- Prevents caching at the Proxy Server -->
        <meta http-equiv="Expires" content="0"/>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
        <meta name="generator" content="AppFuse 2.0" /> <!-- leave this for stats please -->
        <link rel="icon" href="/images/favicon.png" type="image/png"/>
        <title>???BDGAccueil.title??? | X'Glass</title>

		<link rel="stylesheet" type="text/css" href="/scripts/extjs/resources/css/ext-all.css" />
        		
        
		
			<link rel="stylesheet" type="text/css" href="styles/pegase/equipAuto.css" />
	        <l

In [210]:
r = session.get('https://portail-xglass.com/initRechercheVehicule.html')
ensure_logged(r)
print(r.text)

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//FR"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">






















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
    <head>
                <!-- HTTP 1.1 -->
        <meta http-equiv="Cache-Control" content="no-store"/>
        <!-- HTTP 1.0 -->
        <meta http-equiv="Pragma" content="no-cache"/>
        <!-- Prevents caching at the Proxy Server -->
        <meta http-equiv="Expires" content="0"/>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
        <meta name="generator" content="AppFuse 2.0" /> <!-- leave this for stats please -->
        <link rel="icon" href="/images/favicon.png" type="image/png"/>
        <title>???BDGRechercheVehicule.title??? | X'Glass</title>

		<link rel="stylesheet" type="text/css" href="/scripts/extjs/resources/css/ext-all.css" />
		<link rel="stylesheet" type="text/css" href="/scripts/extjs/extensions/resources/Spinner.css" />
    

In [211]:
pprint(r.request.headers.__dict__)

{'_store': OrderedDict([('user-agent',
                         ('user-agent',
                          'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 '
                          '(KHTML, like Gecko) Chrome/67.0.3396.99 '
                          'Safari/537.36')),
                        ('accept-encoding',
                         ('Accept-Encoding', 'gzip, deflate')),
                        ('accept',
                         ('accept',
                          'text/html,application/xhtml+xml,application/xml')),
                        ('connection', ('Connection', 'keep-alive')),
                        ('content-type',
                         ('Content-Type',
                          'application/x-www-form-urlencoded; charset=UTF-8')),
                        ('x-requested-with',
                         ('X-Requested-With', 'XMLHttpRequest')),
                        ('host', ('Host', 'portail-xglass.com')),
                        ('origin', ('Origin', 'https://port

In [212]:
URL = 'https://portail-xglass.com/monActivite.html'
r = session.get(URL,headers={
    "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
    "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36",
    "accept-encoding":"gzip, deflate, br, zstd",
    # "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
    # "X-Requested-With": "XMLHttpRequest",
    # "Host": "portail-xglass.com",
    # "Origin": "https://portail-xglass.com",
    
})
ensure_logged(r)
print(r.text)

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//FR"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">






















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
    <head>
                <!-- HTTP 1.1 -->
        <meta http-equiv="Cache-Control" content="no-store"/>
        <!-- HTTP 1.0 -->
        <meta http-equiv="Pragma" content="no-cache"/>
        <!-- Prevents caching at the Proxy Server -->
        <meta http-equiv="Expires" content="0"/>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
        <meta name="generator" content="AppFuse 2.0" /> <!-- leave this for stats please -->
        <link rel="icon" href="/images/favicon.png" type="image/png"/>
        <title>???BDGMonActivite.title??? | X'Glass</title>

		<link rel="stylesheet" type="text/css" href="/scripts/extjs/resources/css/ext-all.css" />
		<link rel="stylesheet" type="text/css" href="/scripts/extjs/extensions/resources/Spinner.css" />
        
 

In [213]:
URL = "https://portail-xglass.com/ajax/searchImmat.html?selectionsCriteres.immatriculation=DS808DZ&idMarque="
URL = "https://portail-xglass.com/ajax/searchImmat.html"
# r = requests.post(
r = session.post(
    URL,
    # params={
    #    'selectionsCriteres.immatriculation': 'DS808DZ',
    #    'idMarque': '',
    # },
    data={
        'selectionsCriteres.immatriculation': 'DS808DZ',
        'idMarque': '',
    },
    # cookies=session.cookies,
    headers={
        # "Accept": 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7',
        # "Accept-Encoding": "gzip, deflate, br, zstd",
        # "Accept-Language":'en-US,en;q=0.9',
        # "Cache-Control": "no-cache",
        # "Connexion": "keep-alive",
        # 'Host':'portail-xglass.com',
        # "Pragma": "no-cache",
        # "Sec-Ch-Ua": '"Not A(Brand";v="8", "Chromium";v="132", "Google Chrome";v="132"',
        # "Sec-Ch-Ua-Mobile": "?0",
        # "Sec-Ch-Ua-Platform": "Windows",
        # # "Content-Type": "application/json;charset=UTF-8",
        # "Sec-Fetch-Dest": "document",
        # "Sec-Fetch-Mode": "navigate",
    #     "sec-fetch-site": "none",
    #     "sec-fetch-user": "?1",
        # "upgrade-insecure-requests": "1",
    },
    # cookies={
    #     'JSESSIONID':session.cookies.get_dict().get("JSESSIONID"),
    #     # 'SameSite':'None',
    # }
)
ensure_logged(r)
print(r.text)

{"criteresVariante":{"anneesMec":[2012,2013,2014,2015],"boites":[{"label":"196_5","value":"196_5"}],"carrosseries":[{"label":"lbl.carrosserie.203_5","value":"203_5"}],"carrosseriesNbPortesEmpattementHauteur":[{"label":"203_5","value":"203_5"}],"cylindrees":["1.2_null"],"cylindreesCarburant":["1.2_233"],"energies":[{"label":"Essence","value":"233"}],"faps":[0],"marques":[{"codeDarva":"HYUN","codeEtai":"HYU","groupeMarque":{"id":6,"marques":[null,{"codeDarva":"KIA","codeEtai":"KIA","groupeMarque":null,"id":77,"listeDesMarquesExt":[{"application":"DARVA","code":"KIA","id":882,"nom":"KIA"},{"application":"HP","code":null,"id":2341,"nom":"KIA"}],"nom":"KIA","ordre":2}],"nom":"HYUNDAI-KIA","ordre":6},"id":70,"listeDesMarquesExt":[{"application":"HP","code":null,"id":2335,"nom":"HYUNDAI"},{"application":"DARVA","code":"HYUN","id":798,"nom":"HYUNDAI"}],"nom":"HYUNDAI","ordre":2}],"modeles":[{"dateDebut":"2009-04-01T00:00:00","dateFin":"2015-03-31T00:00:00","gamme":"I20","id":2986,"imageRef":"2

In [214]:
pprint(r.json())

{'criteresVariante': {'anneesMec': [2012, 2013, 2014, 2015],
                      'boites': [{'label': '196_5', 'value': '196_5'}],
                      'carrosseries': [{'label': 'lbl.carrosserie.203_5',
                                        'value': '203_5'}],
                      'carrosseriesNbPortesEmpattementHauteur': [{'label': '203_5',
                                                                  'value': '203_5'}],
                      'cylindrees': ['1.2_null'],
                      'cylindreesCarburant': ['1.2_233'],
                      'energies': [{'label': 'Essence', 'value': '233'}],
                      'faps': [0],
                      'marques': [{'codeDarva': 'HYUN',
                                   'codeEtai': 'HYU',
                                   'groupeMarque': {'id': 6,
                                                    'marques': [None,
                                                                {'codeDarva': 'KIA',
                    

In [215]:
session.cookies.get_dict().get("JSESSIONID")

'C26FCC7AF4E37042A69962268FF55F99'

On arrive à obtenir de la data

{"criteresVariante":{"anneesMec":[2012,2013,2014,2015],"boites":[{"label":"196_5","value":"196_5"}],"carrosseries":[{"label":"lbl.carrosserie.203_5","value":"203_5"}],"carrosseriesNbPortesEmpattementHauteur":[{"label":"203_5","value":"203_5"}],"cylindrees":["1.2_null"],"cylindreesCarburant":["1.2_233"],"energies":[{"label":"Essence","value":"233"}],"faps":[0],"marques":[{"codeDarva":"HYUN","codeEtai":"HYU","groupeMarque":{"id":6,"marques":[null,{"codeDarva":"KIA","codeEtai":"KIA","groupeMarque":null,"id":77,"listeDesMarquesExt":[{"application":"DARVA","code":"KIA","id":882,"nom":"KIA"},{"application":"HP","code":null,"id":2341,"nom":"KIA"}],"nom":"KIA","ordre":2}],"nom":"HYUNDAI-KIA","ordre":6},"id":70,"listeDesMarquesExt":[{"application":"HP","code":null,"id":2335,"nom":"HYUNDAI"},{"application":"DARVA","code":"HYUN","id":798,"nom":"HYUNDAI"}],"nom":"HYUNDAI","ordre":2}],"modeles":[{"dateDebut":"2009-04-01T00:00:00","dateFin":"2015-03-31T00:00:00","gamme":"I20","id":2986,"imageRef":"29105.jpg","marque":{"codeDarva":"HYUN","codeEtai":"HYU","groupeMarque":{"id":6,"marques":[null,{"codeDarva":"KIA","codeEtai":"KIA","groupeMarque":null,"id":77,"listeDesMarquesExt":[{"application":"DARVA","code":"KIA","id":882,"nom":"KIA"},{"application":"HP","code":null,"id":2341,"nom":"KIA"}],"nom":"KIA","ordre":2}],"nom":"HYUNDAI-KIA","ordre":6},"id":70,"listeDesMarquesExt":[{"application":"HP","code":null,"id":2335,"nom":"HYUNDAI"},{"application":"DARVA","code":"HYUN","id":798,"nom":"HYUNDAI"}],"nom":"HYUNDAI","ordre":2},"nom":"I20 I (09->15)","ordreGamme":999,"serie":"I"}],"moteurs":[{"label":"G4LA","value":"G4LA"}],"phases":["2"],"puissanceComs":[85],"puissancesFiscales":[4,5]},"errorMessage":null,"infoMessage":null,"regroupementVariantes":[{"listeDeVariante":[{"alimentation":188,"capaciteCarter":3.5,"carrosserie":203,"cylindreeCm3":1248,"cylindreeLit":"1.2","cylindresNbr":4,"dateDebut":"2012-06-01T00:00:00","dateFin":"2015-03-31T00:00:00","detailVariante":[{"flagWildCard":0,"id":25738041,"typeDetail":62,"valeur":"4","variante":null},{"flagWildCard":0,"id":25705123,"typeDetail":62,"valeur":"5","variante":null}],"distributionElement":null,"distributionType":"PIGNONS","energie":233,"essieuMoteur":112,"fap":0,"id":397899,"imageRef":"37226.jpg","injection":null,"libelleCourt":"HYUNDAI I20 I PHASE 2 - 5P 2012-06->2015-03 1.2i 85","modele":{"dateDebut":"2009-04-01T00:00:00","dateFin":"2015-03-31T00:00:00","gamme":"I20","id":2986,"imageRef":"29105.jpg","marque":{"codeDarva":"HYUN","codeEtai":"HYU","groupeMarque":{"id":6,"marques":[null,{"codeDarva":"KIA","codeEtai":"KIA","groupeMarque":null,"id":77,"listeDesMarquesExt":[{"application":"DARVA","code":"KIA","id":882,"nom":"KIA"},{"application":"HP","code":null,"id":2341,"nom":"KIA"}],"nom":"KIA","ordre":2}],"nom":"HYUNDAI-KIA","ordre":6},"id":70,"listeDesMarquesExt":[{"application":"HP","code":null,"id":2335,"nom":"HYUNDAI"},{"application":"DARVA","code":"HYUN","id":798,"nom":"HYUNDAI"}],"nom":"HYUNDAI","ordre":2},"nom":"I20 I (09->15)","ordreGamme":999,"serie":"I"},"modeleCommercial":"I20 I PHASE 2 - 5P 2012-06->2015-03","nom":"1.2i 85","ordre":null,"phase":"2","portesNbr":5,"puissanceCom":85,"puissanceKw":62,"statut":null,"typeBoiteVeitesses":"M5EF2","typeEmpattement":null,"typeHauteur":null,"typeMoteur":"G4LA","typeVehicule":163,"vitessesBte":196,"vitessesNbr":5}],"modeleCommercial":"I20 I PHASE 2 - 5P 2012-06->2015-03"}],"selectionsCriteres":{"anneeDeCommercialisation":null,"carrosserie":null,"criteres":{},"cylindreeLit":null,"energie":null,"idMarque":null,"idModele":null,"identAdmin":null,"immatriculation":"DS808DZ","portesNbr":null,"puissanceFisc":null,"vitessesBte":null,"vitessesNbr":null}}

 en passant le cookie obtenu dans dans une fenêtre de navigation, il faut que l'utilisateur ait quand même chargé la page 'initRecherche' pour que la requête soit acceptée.

In [229]:
varianteId = 397899
r = session.get('https://portail-xglass.com/selectVehicule.html?sessionScopedBean.infoSelectionVehicule.variante.id=%s' %varianteId)
print(r.text)

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//FR"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">






















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
    <head>
                <!-- HTTP 1.1 -->
        <meta http-equiv="Cache-Control" content="no-store"/>
        <!-- HTTP 1.0 -->
        <meta http-equiv="Pragma" content="no-cache"/>
        <!-- Prevents caching at the Proxy Server -->
        <meta http-equiv="Expires" content="0"/>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
        <meta name="generator" content="AppFuse 2.0" /> <!-- leave this for stats please -->
        <link rel="icon" href="/images/favicon.png" type="image/png"/>
        <title>???BDGSelectionPieces.title??? | X'Glass</title>

		<link rel="stylesheet" type="text/css" href="/scripts/extjs/resources/css/ext-all.css" />
		<link rel="stylesheet" type="text/css" href="/scripts/extjs/extensions/resources/Spinner.css" />
      

In [224]:
page = bs.BeautifulSoup(r.text, "html.parser")

In [239]:
scripts = page.find_all("script",src=False)
scripts

[<script type="text/javascript">
 			function createPopupToDisplayContent(contentUrl, popupTitle) {
 				new PopupDisplayContent({'contentUrl' : contentUrl, 'title' : popupTitle}).show();
 			}
 			
 			function getIdDevisActif (){
 		 		// Récupère l'identifiant du devis passé par paramètre URL
 				var idDevisActif = '';
 		 		if(idDevisActif != '') {return idDevisActif;}
 		 		// Sinon récupère l'identifiant du devis passé par paramètre JSON
 		 		idDevisActif = '';
 		 		if(idDevisActif != '') {return idDevisActif;}
 		 		// Sinon récupère l'identifiant du devis actif stocké dans le dossier
 		 		idDevisActif = '';
 		 		return idDevisActif;
 		 	}
 			
 			// Le namespace Ext.pegase est déclaré dans global.js
 			Ext.apply(Ext.pegase, {
 				version : '3.19.4',
 				versionDB : '3.29.1',
 				contextPath : '/',
 				cgvUrl : 'CGV/21/BDG_DEVIS_ASSURE/CGV_26346.pdf',
 				defaultCgvUrl : '/CGV/CGV_default.pdf',
 				cgvPopupTitle : Ext.I18n.getLabel('documentation.lbl.cgv'),
 				ne

In [252]:
"planche =" in scripts[3].text

True

In [274]:
planche = None
for script in scripts:
    # print(script.text)
    # print("planche =" in script.text)
    try:
        if "planche =" in script.text:
            # print(scripts.text)
            lines = script.text.splitlines()
            # print(lines)
            for line in lines:
                # print(line)
                if "planche =" in line:
                    raw = line.split("planche = ")[1]
                    raw = raw.replace(";", "")
                    print(raw)
                    # planche = 
    except:
        pass
        
pprint(planche)

{"calques":[{"codification":"fam2207G","id":68,"libelle":"FEU REPETITEUR LATERAL","typeCalque":""},{"codification":"fam3205","id":284,"libelle":"MECANISME ESSUIE- GLACE AR","typeCalque":""},{"codification":"fam2114","id":54,"libelle":"ESSUIE-GLACE AR","typeCalque":""},{"codification":"fam2115","id":57,"libelle":"ESSUIE-GLACE AV","typeCalque":""},{"codification":"fam3016","id":55,"libelle":"JT/LECHEUR PORTE AR","typeCalque":""},{"codification":"fam2453D","id":240,"libelle":"GLACE RETROVISEUR","typeCalque":""},{"codification":"fam2202D","id":69,"libelle":"FEU AR","typeCalque":""},{"codification":"fam4209G","id":71,"libelle":"RETROVISEUR EXT","typeCalque":""},{"codification":"fam3545","id":67,"libelle":"PHARE AB/LP","typeCalque":""},{"codification":"fam6030","id":252,"libelle":"GLACE DE PAVILLON","typeCalque":""},{"codification":"fam2449D","id":47,"libelle":"GLACE PORTE AR","typeCalque":""},{"codification":"fam3204","id":285,"libelle":"MECANISME ESSUIE- GLACE","typeCalque":""},{"codificat

In [278]:
import json
json.loads(raw)

{'calques': [{'codification': 'fam2207G',
   'id': 68,
   'libelle': 'FEU REPETITEUR LATERAL',
   'typeCalque': ''},
  {'codification': 'fam3205',
   'id': 284,
   'libelle': 'MECANISME ESSUIE- GLACE AR',
   'typeCalque': ''},
  {'codification': 'fam2114',
   'id': 54,
   'libelle': 'ESSUIE-GLACE AR',
   'typeCalque': ''},
  {'codification': 'fam2115',
   'id': 57,
   'libelle': 'ESSUIE-GLACE AV',
   'typeCalque': ''},
  {'codification': 'fam3016',
   'id': 55,
   'libelle': 'JT/LECHEUR PORTE AR',
   'typeCalque': ''},
  {'codification': 'fam2453D',
   'id': 240,
   'libelle': 'GLACE RETROVISEUR',
   'typeCalque': ''},
  {'codification': 'fam2202D',
   'id': 69,
   'libelle': 'FEU AR',
   'typeCalque': ''},
  {'codification': 'fam4209G',
   'id': 71,
   'libelle': 'RETROVISEUR EXT',
   'typeCalque': ''},
  {'codification': 'fam3545',
   'id': 67,
   'libelle': 'PHARE AB/LP',
   'typeCalque': ''},
  {'codification': 'fam6030',
   'id': 252,
   'libelle': 'GLACE DE PAVILLON',
   'typeCal

In [279]:
LOGOUT_URL = 'https://portail-xglass.com/logout.html'
r = session.get(LOGOUT_URL)
print(r.text)




















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
<head>
    <title>X'Glass - Le Portail Bris de Glace </title>
	<!-- meta dediee a le redirection lors des appels ajax en etant deconnecte -->
	<meta name="xglass-ajax-redirect" />

	<meta name="description" content="X'Glass est la solution de chiffrage pour les pièces de vitrage et optiques pour les véhicules de type VL et VUL."/>


	<link rel="stylesheet" type="text/css" href="scripts/extjs/resources/css/ext-all.css" />
    
    
    
		<link rel="stylesheet" type="text/css" href="styles/pegase/equipAuto.css" />
	    <link rel="stylesheet" type="text/css" href="/styles/pegase/theme.css" />
	    
	    <script type="text/javascript" src="/scripts/prototype.js"></script>
	    <script type="text/javascript" src="/scripts/scriptaculous.js"></script>
	    <script type="text/javascript" src="/scripts/extjs/adapter/ext/ext-base.js"></script>
		<script type="text/javascript" src="/scripts/extjs/ext-all-debug.js"></scr

In [4]:
from xglass import *
from pprint import pprint
xglass = XGLASS()
r = xglass.auth()
try:
    # xglass.close()
    vehicules = xglass.searchVehiculeImmat('EZ817MG')
    vehicule = vehicules[0]
    # pprint(vehicule.__dict__)
    planche = xglass.getPlancheData(vehicule)
    pare_brise = [c for c in planche.xGlassCalques if c.libelle == "PARE-BRISE"][0]
    # print(pare_brise.__dict__)
    # r = xglass.affichagePieces(planche, pare_brise)
    xglass.getPieces(planche, pare_brise)
    pare_briseElement = xglass.elements_principaux[0]
    for pare_brise_piece_avecAM in pare_briseElement.pieces:
        pprint(pare_brise_piece_avecAM.eurocode)

    pare_brise_piece_avecAM = [piece for piece in pare_briseElement.pieces if piece.eurocode == '861101J110'][0]
    pare_brise_piece_sansAM = [piece for piece in pare_briseElement.pieces if piece.eurocode == '861104P010'][0]

    pare_brise_piece_avecAMPieces = xglass.getPieceAm(pare_briseElement, pare_brise_piece_avecAM)
    pare_brise_piece_sansAMPieces = xglass.getPieceAm(pare_briseElement, pare_brise_piece_sansAM)
    # pare_brise_AMPieces = xglass.getPieceAm(pare_briseElement)
except Exception as e:
    print(e)
# xglass.close()



xglass.close()

'84051M68P02-000'
list index out of range
Logout success


In [3]:
a = pare_brise_piece_avecAMPieces[0]
a.id

12298535

In [16]:
pprint(pare_brise_piece_avecAMPieces[0].__dict__)

{'debutValidite': '2009',
 'eurocode': '4031802',
 'finValidite': None,
 'id': 12298535,
 'pieceAm': {'caracteristiquesPieceValeur': None,
             'criterePieces': [{'ciceron': False,
                                'cleTradValeur': 'a0d6b8b8b6ff239253dbc1af10337dabd12f2236',
                                'id': 5866,
                                'type': {'cicerone': False,
                                         'cleTradLibelle': '248585f8774519ddd87cfead61930eb738f3ba21',
                                         'discriminantArbo': False,
                                         'discriminantFiltre': False,
                                         'id': 2019,
                                         'image': None,
                                         'libelle': 'Couleur vitrage',
                                         'ordre': None,
                                         'tolerance': None,
                                         'traduisible': True,
               

In [2]:
pare_brise_piece_avecAMPieces

In [9]:
r = xglass.findSelectionsPiecesAmView(pare_briseElement, pare_brise_piece_sansAM)
pprint(r.json())

{'selectionsPiecesAmView': []}


In [10]:
r.json()['selectionsPiecesAmView']

[]

In [ ]:
r = xglass.findSelectionsPiecesAmView(pare_briseElement, pare_brise_piece_avecAM)
pprint(r.json())

{'selectionsPiecesAmView': [{'debutValidite': '2009',
                             'finValidite': None,
                             'pieceAm': {'caracteristiquesPieceValeur': None,
                                         'criterePieces': [{'ciceron': False,
                                                            'cleTradValeur': 'a0d6b8b8b6ff239253dbc1af10337dabd12f2236',
                                                            'id': 5866,
                                                            'type': {'cicerone': False,
                                                                     'cleTradLibelle': '248585f8774519ddd87cfead61930eb738f3ba21',
                                                                     'discriminantArbo': False,
                                                                     'discriminantFiltre': False,
                                                                     'id': 2019,
                                                     

In [6]:
r = xglass.findSelectionsPiecesAmView(pare_briseElement)
pprint(r.json())

{'selectionsPiecesAmView': [{'debutValidite': '2009',
                             'finValidite': None,
                             'pieceAm': {'caracteristiquesPieceValeur': None,
                                         'criterePieces': [{'ciceron': False,
                                                            'cleTradValeur': '4cf59c52bb7ce29029b03da14a6d480d715d7afb',
                                                            'id': 14787078,
                                                            'type': {'cicerone': False,
                                                                     'cleTradLibelle': '03ed996b6e4ffa63b3237524dff605b75a17d062',
                                                                     'discriminantArbo': False,
                                                                     'discriminantFiltre': False,
                                                                     'id': 88295,
                                                

In [2]:
xglass.close()

Logout success


In [ ]:
pare_briseElement = xglass.elements_principaux[0]
for pare_brise_piece_avecAM in pare_briseElement.pieces:
    pprint(pare_brise_piece_avecAM.eurocode)
    # xglass.getPiece(piece)

'861104P010'
'861104P210'
'861101J110'
'861104P310'


In [ ]:
pare_brise_piece_avecAM = [piece for piece in pare_briseElement.pieces if piece.eurocode == '861101J110'][0]
pare_brise_piece_avecAM.id

530828

In [ ]:
from pprint import pprint
for pare_brise_piece_avecAM in xglass.elements_principaux: pprint(pare_brise_piece_avecAM.__dict__)

{'addForfaitOptique': False,
 'affichageAm': True,
 'containsPiecesAm': True,
 'elementSitId': '3464',
 'elementSitLibelle': 'PARE-BRISE',
 'elementSitTypeIsOptique': False,
 'elementVitre': True,
 'elementVitreMobile': False,
 'forfaitKitCollage': False,
 'libelle': 'PARE-BRISE',
 'listPieceAmView': [],
 'listPieceOeView': [{'containsPiecesAm': False,
                      'filterPieceByVin': False,
                      'pieceOe': {'caracteristiques': [{'codearcauto': '185002',
                                                        'id': 4,
                                                        'ordre': 2,
                                                        'type': {'code': 'PB_CH',
                                                                 'id': 4,
                                                                 'libelle': 'Chauffant '
                                                                            '(PB)',
                                                     

In [6]:
xglass

In [ ]:
[pprint(piece.__dict__) for piece in xglass.elements_complementaires] 

{'addForfaitOptique': False,
 'affichageAm': True,
 'containsPiecesAm': True,
 'elementSitId': '1922',
 'elementSitLibelle': 'EMBASE RETROVISEUR INTERIEUR',
 'elementSitTypeIsOptique': False,
 'elementVitre': False,
 'elementVitreMobile': False,
 'forfaitKitCollage': False,
 'libelle': 'EMBASE RETROVISEUR INTERIEUR',
 'listPieceAmView': [],
 'listPieceOeView': [{'containsPiecesAm': False,
                      'filterPieceByVin': False,
                      'pieceOe': {'caracteristiques': [],
                                  'codearcauto': '187999',
                                  'couleur': '',
                                  'datePrix': {'date': 1,
                                               'day': 3,
                                               'hours': 0,
                                               'minutes': 0,
                                               'month': 0,
                                               'nanos': 0,
                                        

[None, None, None, None]

In [19]:
r = xglass.session.post('https://portail-xglass.com/affichagePieces.html?planche.id=26881&calque.id=59&filtrageVinButtonClickedForDevisRapide=true')

In [20]:
print(r.text)

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//FR"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">






















<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr">
    <head>
                <!-- HTTP 1.1 -->
        <meta http-equiv="Cache-Control" content="no-store"/>
        <!-- HTTP 1.0 -->
        <meta http-equiv="Pragma" content="no-cache"/>
        <!-- Prevents caching at the Proxy Server -->
        <meta http-equiv="Expires" content="0"/>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
        <meta name="generator" content="AppFuse 2.0" /> <!-- leave this for stats please -->
        <link rel="icon" href="/images/favicon.png" type="image/png"/>
        <title>???BDGAffichagePieces.title??? | X'Glass</title>

		<link rel="stylesheet" type="text/css" href="/scripts/extjs/resources/css/ext-all.css" />
		<link rel="stylesheet" type="text/css" href="/scripts/extjs/extensions/resources/Spinner.css" />
      

In [11]:
planche.id

26881

In [12]:
pare_brise.id

59

In [21]:
print(r.request.path_url)

/affichagePieces.html?planche.id=26881&calque.id=59&filtrageVinButtonClickedForDevisRapide=true


In [22]:
page = bs.BeautifulSoup(r.text, "html.parser")
scripts = page.find_all("script",src=False)
for script in scripts:
    try:
        if "var elementSitMapData =" in script.text:
            lines = script.text.splitlines()
    except:
        pass
for line in lines:
    print(line)


<!--
	var erreur = '';
	var elementSitMapData = [{"ELEMENTSIT_PRINCIPAUX":[{"addForfaitOptique":false,"affichageAm":true,"containsPiecesAm":true,"elementSitId":"3464","elementSitLibelle":"PARE-BRISE","elementSitTypeIsOptique":false,"elementVitre":true,"elementVitreMobile":false,"forfaitKitCollage":false,"listPieceAmView":[],"listPieceOeView":[{"containsPiecesAm":false,"filterPieceByVin":false,"pieceOe":{"caracteristiques":[{"codearcauto":"185002","id":4,"ordre":2,"type":{"code":"PB_CH","id":4,"libelle":"Chauffant (PB)","ordre":4},"valeur":"false"},{"codearcauto":"185002","id":5,"ordre":2,"type":{"code":"PB_AN","id":7,"libelle":"Antenne (PB)","ordre":7},"valeur":"false"},{"codearcauto":"185002","id":3,"ordre":1,"type":{"code":"PB_CA","id":2,"libelle":"Capteur (PB)","ordre":2},"valeur":"sans"},{"codearcauto":"185002","id":7,"ordre":2,"type":{"code":"PB_B","id":5,"libelle":"Bandeau (PB)","ordre":5},"valeur":"false"},{"codearcauto":"185002","id":6,"ordre":2,"type":{"code":"PB_D","id":6,"l

In [23]:
import json
for line in lines:
    if "var elementSitMapData =" in line:
        raw = line.split("var elementSitMapData = ")[1]
        raw = raw.replace(";", "")
        # print(raw)
        # planche =
        data = json.loads(raw)
        break

data

[{'ELEMENTSIT_PRINCIPAUX': [{'addForfaitOptique': False,
    'affichageAm': True,
    'containsPiecesAm': True,
    'elementSitId': '3464',
    'elementSitLibelle': 'PARE-BRISE',
    'elementSitTypeIsOptique': False,
    'elementVitre': True,
    'elementVitreMobile': False,
    'forfaitKitCollage': False,
    'listPieceAmView': [],
    'listPieceOeView': [{'containsPiecesAm': False,
      'filterPieceByVin': False,
      'pieceOe': {'caracteristiques': [{'codearcauto': '185002',
         'id': 4,
         'ordre': 2,
         'type': {'code': 'PB_CH',
          'id': 4,
          'libelle': 'Chauffant (PB)',
          'ordre': 4},
         'valeur': 'false'},
        {'codearcauto': '185002',
         'id': 5,
         'ordre': 2,
         'type': {'code': 'PB_AN',
          'id': 7,
          'libelle': 'Antenne (PB)',
          'ordre': 7},
         'valeur': 'false'},
        {'codearcauto': '185002',
         'id': 3,
         'ordre': 1,
         'type': {'code': 'PB_CA',
       

In [ ]:
for pare_brise_piece_avecAM in xglass.elements_principaux[0].pieces:
    eurocode = pare_brise_piece_avecAM.pieceOe.get('reference')
    if eurocode == '861101J110':
        break
pprint(pare_brise_piece_avecAM.__dict__)


{'containsPiecesAm': False,
 'filterPieceByVin': False,
 'pieceOe': {'caracteristiques': [{'codearcauto': '185003',
                                   'id': 178,
                                   'ordre': 2,
                                   'type': {'code': 'PB_T',
                                            'id': 1,
                                            'libelle': 'Teinte (PB)',
                                            'ordre': 1},
                                   'valeur': 'teinte'},
                                  {'codearcauto': '185003',
                                   'id': 181,
                                   'ordre': 2,
                                   'type': {'code': 'PB_CH',
                                            'id': 4,
                                            'libelle': 'Chauffant (PB)',
                                            'ordre': 4},
                                   'valeur': 'false'},
                                  {'codearc

In [ ]:
def findSelectionsPiecesAmView(self, element:XGlassElement, piece:XGlassPiece = None):
    URL = 'https://portail-xglass.com/ajax/findSelectionsPiecesAmView.html'
    data = {
        'withPiecesAm':element.withPiecesAm,
        'idDevis':''
    }
    if piece:
        data['idPieceOe'] = piece.id
    else:
        data['idElementSit'] = element.elementSitId

    r = self.session.post(
        URL,
        data=data,
    )
    return r